Develop2

ebatuev opened the pull request:
Создал отдельный репозиторий для тестов API.
 Создал Maven-проект.
 Подключил JUnit 4, RestAssured и Allure.

 Проект:
1. Cоздать класс PathAPI с эндпоинтами (ручками).

2. Cоздать классы для пользователя UserRegistration, UserLogin, UserChangeData и для заказа OrderCreate.
* Начать описывать - Сериализацию — преобразовать объект в JSON
*
* Как использовать сериализацию, если нужно отправить запрос с JSON:
*  1) Создать класс, поля которого соответствуют ключам JSON. Добавить конструктор с параметрами и конструктор без параметров.
*  2) В тесте создать объект класса с нужными значениями — через конструктор.
*  3) Сериализовать объект, то есть преобразовать его в JSON. Для этого нужно подключить библиотеку Gson.

3. Описать общие шаги для пользователя и заказа и вынести их в классы User и Order.
   Классы являются наследниками класса PathAPI, где хранятся ручки.

4. Создали класс BaseMethods с переменными, аннотацией @Before (url stellarburgers) и @After (удаление пользователя), геттерами.

5. Пишем тестовые классы, начиная с класса регистрации пользователя RegistrationUserTest,
   где например:
   - отправляем запросы и сохраняет ответ в переменную response, экзмепляр класса Response;
   - проверяем, что в теле ответа ключам соответствуют правильные значения, а Матчер notNullValue() проверяет, что аргумент метода assertThat — не null-значение;
   - проверяем, что статус-код ответа верен.
   По такому же принципу создаем тесты для классов LoginUserTest, ChangeUserDataTest и для заказов CreateOrderTest и UserOrdersTest.

6. Создаем тест с параметризацией RegistrationUserAndDoNotFillInOneOfRequiredFields, где:
   - прописали аннотации
   - создали поля тестового класса
   - объявили конструктор класса
   - сам тест

7. Сделал отчёт в Allure:
   - выполнил команду mvn clean test, чтобы сгенерировать отчёт
   - в консоли перейшел в папку проекта и выполни команду allure serve target/surefire-reports/. Allure отобразит отчёт, который сгенерировал плагин maven-surefire-plugin

Commit:
1 коммит - API

Commit:
2 коммит - Allure

Commit:
3 коммит - Для прохождения теста createOrderWithoutAuthorization должен приходить ответ 401,но приходит 200 => в тесте оставил 200, чтобы тест не падал. Это баг системы, который найден автотестом.

nikolay-977 commented on src/main/java/model_POJO/OrderCreate.java:
```diff
@@ -0,0 +1,30 @@
+package model_POJO;
```
Необходимо исправить наименование пакетов в соответствии с правилами нейминга. Пакеты — допустимы только строчные буквы. В случае, если слов несколько, их следует разделять точкой. Префикс должен быть одним из доменных имен верхнего уровня, таких как com, ru и т.д. Наименование пакета должно отражать суть содержимого.

ebatuev replied:
Исправлен нейминг пакетов в соответствии с правилами

nikolay-977 commented on src/test/java/ChangeUserDataTest.java:
```diff
@@ -0,0 +1,41 @@
+        String accessToken = responseCreate.jsonPath().get("accessToken");
+        Response responseChange = getUser().changeUserData(getUserChangeData(), accessToken);
+        responseChange.then().assertThat()
+                .body("success", equalTo(true))
```
Необходимо исправить порядок проверки - сначала проверяем statusCode, затем остальное. Иначе в случае если сервер недоступен мы получим NullPointerException, а не status сode 500-й.

ebatuev replied:
Важное замечание, спасибо)
исправлен порядок проверки - сначала проверяем statusCode, затем остальное.

nikolay-977 commented on src/test/java/RegistrationUserAndDoNotFillInOneOfRequiredFields.java:
```diff
@@ -0,0 +1,49 @@
+        this.name = name;
+    }
+
+    @Parameterized.Parameters // добавили аннотацию
```
Можно улучшить: В параметризованных тестах для аннотации `@Parameterized.Parameters` лучше использовать аргумент name: `@Parameterized.Parameters(name = "Стоимость булочки. Тестовые данные: {0} {1}")`, где {0}, {1} - индексы параметров. С аргументом name наименование теста отобразится в отчете. А использование индексов параметров повысит информативность проверки.

ebatuev replied:
Добавлен аргумент name в параметризованном тесте для аннотации.

nikolay-977 commented on src/main/java/step/BaseMethods.java:
```diff
@@ -0,0 +1,110 @@
+
+
+
+    private final UserRegistration userRegistration = new UserRegistration(email, password, name);
```
Необходимо исправить: создание пользователя и получение токена следует выполнять в методе с аннотацией `@Before`.

ebatuev replied:
Поправил аннотацию @Before

Commit:
3 коммит -
1) Исправлен нейминг пакетов в соответствии с правилами
2) Исправлен порядок проверки - сначала проверяем statusCode, затем остальное
3) Добавлен аргумент name в параметризованном тесте для аннотации
4) Поправил аннотацию

Commit:
4 коммит - добавлен отчёт Allure
